File: highspeed_report/highspeed_accounts/report/executive_daily_sales_summary/executive_daily_sales_summary.py
# Copyright (c) 2026, Highspeed and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, add_days, getdate

def execute(filters=None):
    if not filters:
        filters = {}

    columns = get_columns()
    data = get_data(filters)
    summary = get_report_summary(data)

    return columns, data, None, None, summary

def get_columns():
    return [
        {
            "label": _("Particulars"),
            "fieldname": "metric",
            "fieldtype": "Data",
            "width": 260
        },
        {
            "label": _("Today's Value"),
            "fieldname": "today_val",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Yesterday's Value"),
            "fieldname": "yesterday_val",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Change"),
            "fieldname": "change_val",
            "fieldtype": "Data",
            "width": 130
        },
        {
            "label": _("Change %"),
            "fieldname": "change_percent",
            "fieldtype": "Data",
            "width": 110
        }
    ]

def get_data(filters):
    company = filters.get("company")
    target_date = getdate(filters.get("posting_date") or frappe.utils.today())
    yesterday_date = add_days(target_date, -1)

    # 1. Helper function to query sales summary for a date
    def query_sales_summary(date_str):
        conditions = ["si.docstatus = 1", "si.posting_date = %(date)s"]
        values = {"date": date_str}
        if company:
            conditions.append("si.company = %(company)s")
            values["company"] = company
            
        # Sales query
        sales_data = frappe.db.sql(f"""
            SELECT 
                COALESCE(SUM(base_grand_total), 0) as total_sales,
                COUNT(name) as invoice_count
            FROM `tabSales Invoice` si
            WHERE {" AND ".join(conditions)} AND is_return = 0
        """, values, as_dict=True)[0]

        # Returns query
        returns_data = frappe.db.sql(f"""
            SELECT 
                ABS(COALESCE(SUM(base_grand_total), 0)) as total_returns
            FROM `tabSales Invoice` si
            WHERE {" AND ".join(conditions)} AND is_return = 1
        """, values, as_dict=True)[0]

        sales = flt(sales_data.total_sales)
        count = flt(sales_data.invoice_count)
        returns = flt(returns_data.total_returns)
        basket = (sales / count) if count else 0.0
        net = sales - returns

        return {
            "sales": sales,
            "count": count,
            "basket": basket,
            "returns": returns,
            "net_sales": net
        }

    t_data = query_sales_summary(target_date)
    y_data = query_sales_summary(yesterday_date)

    # Calculate metrics row helper
    def get_row(metric_name, t_val, y_val, is_currency=True):
        change = t_val - y_val
        change_pct = (change / y_val * 100) if y_val else (100.0 if t_val else 0.0)
        
        currency_sym = frappe.get_cached_value("Company", company, "default_currency") if company else "SAR"
        
        def format_val(val):
            if is_currency:
                return frappe.utils.fmt_money(val, currency=currency_sym)
            return str(int(val)) if val.is_integer() else f"{val:.2f}"

        return {
            "metric": metric_name,
            "today_val": format_val(t_val),
            "yesterday_val": format_val(y_val),
            "change_val": format_val(change),
            "change_percent": f"{change_pct:+.1f}%" if change_pct else "0.0%",
            "raw_change_pct": change_pct
        }

    data = []

    # Section 1 Header
    data.append({
        "metric": "--- Executive Summary ---",
        "today_val": "",
        "yesterday_val": "",
        "change_val": "",
        "change_percent": ""
    })

    data.append(get_row(_("Total Sales"), t_data["sales"], y_data["sales"]))
    data.append(get_row(_("Total Invoices"), t_data["count"], y_data["count"], is_currency=False))
    data.append(get_row(_("Average Basket Value"), t_data["basket"], y_data["basket"]))
    data.append(get_row(_("Total Returns"), t_data["returns"], y_data["returns"]))
    data.append(get_row(_("Net Sales"), t_data["net_sales"], y_data["net_sales"]))

    # Section 2 Header: Top 5 Selling Items
    data.append({
        "metric": "--- Top 5 Selling Items (Today) ---",
        "today_val": "",
        "yesterday_val": "",
        "change_val": "",
        "change_percent": ""
    })

    # Query top items
    top_items_conditions = ["si.docstatus = 1", "si.posting_date = %(date)s", "si.is_return = 0"]
    top_items_values = {"date": target_date}
    if company:
        top_items_conditions.append("si.company = %(company)s")
        top_items_values["company"] = company

    top_items = frappe.db.sql(f"""
        SELECT 
            sii.item_code,
            sii.item_name,
            SUM(sii.qty) as qty,
            SUM(sii.base_net_amount) as amount
        FROM 
            `tabSales Invoice Item` sii
        INNER JOIN 
            `tabSales Invoice` si ON sii.parent = si.name
        WHERE 
            {" AND ".join(top_items_conditions)}
        GROUP BY 
            sii.item_code, sii.item_name
        ORDER BY 
            qty DESC
        LIMIT 5
    """, top_items_values, as_dict=True)

    currency_sym = frappe.get_cached_value("Company", company, "default_currency") if company else "SAR"

    for idx, item in enumerate(top_items):
        item_label = f"#{idx+1} - {item.item_code} ({item.item_name})"
        qty_formatted = f"{flt(item.qty):.2f}"
        amt_formatted = frappe.utils.fmt_money(item.amount, currency=currency_sym)
        
        data.append({
            "metric": item_label,
            "today_val": qty_formatted,
            "yesterday_val": amt_formatted,
            "change_val": _("Qty Sold"),
            "change_percent": _("Sales Amount")
        })

    return data

def get_report_summary(data):
    # Retrieve Net Sales row
    net_sales_row = next((r for r in data if r["metric"] == _("Net Sales")), None)
    total_invoices_row = next((r for r in data if r["metric"] == _("Total Invoices")), None)
    basket_row = next((r for r in data if r["metric"] == _("Average Basket Value")), None)
    
    summary = []
    if net_sales_row:
        summary.append({
            "value": net_sales_row["today_val"],
            "label": _("Net Sales (Today)"),
            "indicator": "green",
            "datatype": "Data"
        })
    if total_invoices_row:
        summary.append({
            "value": total_invoices_row["today_val"],
            "label": _("Total Invoices (Today)"),
            "indicator": "blue",
            "datatype": "Data"
        })
    if basket_row:
        summary.append({
            "value": basket_row["today_val"],
            "label": _("Average Basket (Today)"),
            "indicator": "orange",
            "datatype": "Data"
        })
        
    return summary

File: highspeed_report/highspeed_accounts/report/executive_daily_sales_summary/executive_daily_sales_summary.js
// Copyright (c) 2026, Highspeed and contributors
// For license information, please see license.txt

frappe.query_reports["Executive Daily Sales Summary"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1,
            "width": "200px"
        },
        {
            "fieldname": "posting_date",
            "label": __("Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1,
            "width": "120px"
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        if (!data) {
            return default_formatter(value, row, column, data);
        }

        if (column.fieldname === "metric" && value) {
            if (value.startsWith("---")) {
                var clean_val = value.replace(/---/g, '').trim();
                return '<span style="font-weight: bold; font-size: 1.15em; color: #2c3e50; background-color: #f1f5f9; padding: 6px 12px; display: block; border-radius: 4px; border-inline-start: 4px solid #475569;">' + 
                        __(clean_val) + 
                       '</span>';
            }
            return __(value);
        }

        // Color coding for KPI changes
        if (column.fieldname === "change_percent" && value) {
            if (value.startsWith("+")) {
                return `<span style="color: #27ae60; font-weight: bold;">▲ ${value}</span>`;
            } else if (value.startsWith("-")) {
                return `<span style="color: #e74c3c; font-weight: bold;">▼ ${value}</span>`;
            }
        }

        if (column.fieldname === "change_val" && value) {
            if (value.startsWith("+")) {
                return `<span style="color: #27ae60; font-weight: bold;">${value}</span>`;
            } else if (value.startsWith("-")) {
                return `<span style="color: #e74c3c; font-weight: bold;">${value}</span>`;
            }
        }

        // For top selling items, style helper text labels
        if (data.metric && data.metric.startsWith("#")) {
            if (column.fieldname === "change_val" || column.fieldname === "change_percent") {
                return `<span style="color: #7f8c8d; background-color: #f5f6fa; padding: 2px 6px; border-radius: 4px; font-size: 11px; font-weight: bold;">${__(value)}</span>`;
            }
            if (column.fieldname === "today_val") {
                return `<span style="font-weight: bold; color: #2980b9;">${value}</span>`;
            }
            if (column.fieldname === "yesterday_val") {
                return `<span style="font-weight: bold; color: #27ae60;">${value}</span>`;
            }
        }

        return default_formatter(value, row, column, data);
    },
    "onload": function(report) {
        report.page.add_inner_button(__("طباعة التقرير"), function() {
            if (report.data && report.data.length > 0) {
                const html = get_executive_print_html(report.data, report.get_values());
                var w = window.open();
                if (w) {
                    w.document.write(html);
                    w.document.close();
                    setTimeout(function() {
                        w.print();
                    }, 500);
                } else {
                    frappe.msgprint(__("تم حظر النوافذ المنبثقة. يرجى السماح بالنوافذ المنبثقة لطباعة التقرير."));
                }
            } else {
                frappe.msgprint(__("لا توجد بيانات لعرضها."));
            }
        });
    }
};

function get_executive_print_html(data, filters) {
    let kpi_rows_html = '';
    let top_items_rows_html = '';
    
    let current_section = ''; // 'kpi' or 'items'
    
    data.forEach(row => {
        if (row.metric && row.metric.startsWith("---")) {
            if (row.metric.includes("Executive Summary")) {
                current_section = 'kpi';
            } else if (row.metric.includes("Top 5 Selling Items")) {
                current_section = 'items';
            }
            return;
        }
        
        if (current_section === 'kpi') {
            let change_color = '#2c3e50';
            if (row.change_percent.startsWith("+")) {
                change_color = '#27ae60';
            } else if (row.change_percent.startsWith("-")) {
                change_color = '#e74c3c';
            }
            
            kpi_rows_html += `
                <tr>
                    <td style="font-weight: bold; text-align: right;">${__(row.metric)}</td>
                    <td class="en-number" style="font-weight: bold; color: #2980b9;">${row.today_val}</td>
                    <td class="en-number" style="color: #7f8c8d;">${row.yesterday_val}</td>
                    <td class="en-number" style="color: ${change_color};">${row.change_val}</td>
                    <td class="en-number" style="font-weight: bold; color: ${change_color};">
                        ${row.change_percent.startsWith("+") ? '▲' : (row.change_percent.startsWith("-") ? '▼' : '')} ${row.change_percent}
                    </td>
                </tr>
            `;
        } else if (current_section === 'items') {
            top_items_rows_html += `
                <tr>
                    <td style="text-align: right; font-weight: bold;">${row.metric}</td>
                    <td class="en-number" style="font-weight: bold; color: #2980b9;">${row.today_val}</td>
                    <td class="en-number" style="font-weight: bold; color: #27ae60;">${row.yesterday_val}</td>
                </tr>
            `;
        }
    });

    const currentDate = frappe.datetime.get_today();
    return `
    <!DOCTYPE html>
    <html dir="rtl">
    <head>
        <meta charset="UTF-8">
        <title>ملخص المبيعات اليومي للإدارة</title>
        <style>
            @page { size: A4 portrait; margin: 0.5cm; }
            body { font-family: Arial, sans-serif; margin: 0; padding: 10px; direction: rtl; font-size: 11px; }
            .header-table { width: 100%; border-collapse: collapse; margin-bottom: 15px; }
            .header-table td { border: none; text-align: center; }
            .header-table h3 { margin: 5px 0; font-size: 16px; }
            .header-table h4 { margin: 5px 0; font-size: 14px; color: #555; }
            .info-table, .data-table { width: 100%; border-collapse: collapse; margin-bottom: 20px; }
            .info-table td, .info-table th { border: 1px solid #ddd; padding: 5px; text-align: right; }
            .data-table td, .data-table th { border: 1px solid #ccc; padding: 8px 6px; text-align: center; }
            .data-table th { background-color: #f5f5f5; font-weight: bold; }
            .section-title { font-size: 13px; font-weight: bold; color: #2c3e50; margin: 15px 0 8px 0; padding-bottom: 3px; border-bottom: 2px solid #34495e; }
            .en-number { font-family: Arial, sans-serif !important; }
        </style>
    </head>
    <body>
        <div class="header-table">
            <table style="width: 100%; border: none;">
                <tr>
                    <td style="text-align: center;">
                        <h3>${filters.company || ""}</h3>
                        <h4>ملخص المبيعات اليومي للإدارة / Executive Daily Sales Summary</h4>
                    </td>
                </tr>
            </table>
        </div>
        <table class="info-table">
            <tr>
                <td style="font-weight: bold; width: 15%;">تاريخ التقرير / Date:</td>
                <td class="en-number" style="text-align: right;">${filters.posting_date || ""}</td>
                <td style="font-weight: bold; width: 15%;">تاريخ الطباعة:</td>
                <td class="en-number" style="text-align: right;">${currentDate}</td>
            </tr>
        </table>
        
        <div class="section-title">الملخص التنفيذي / Executive Summary</div>
        <table class="data-table">
            <thead>
                <tr>
                    <th>البيان / Particulars</th>
                    <th>اليوم / Today</th>
                    <th>الأمس / Yesterday</th>
                    <th>التغير / Change</th>
                    <th>نسبة التغير / Change %</th>
                </tr>
            </thead>
            <tbody>
                ${kpi_rows_html}
            </tbody>
        </table>

        <div class="section-title">أفضل 5 أصناف مبيعاً (اليوم) / Top 5 Selling Items (Today)</div>
        <table class="data-table">
            <thead>
                <tr>
                    <th>الصنف / Item</th>
                    <th>الكمية المباعة / Qty Sold</th>
                    <th>قيمة المبيعات / Sales Amount</th>
                </tr>
            </thead>
            <tbody>
                ${top_items_rows_html}
            </tbody>
        </table>
    </body>
    </html>
    `;
}
